weather_mv/loader_pipeline/ee.py

```diff
@@ -0,0 +1,470 @@
+                                  count=len(data),
+                                  crs=crs,
+                                  transform=transform,
+                                  compress='lzw') as f:
+                    for i, da in enumerate(data):
+                        f.write(da, i+1)
+
+                # Copy in-memory tiff to gcs.
+                target_path = os.path.join(self.tiff_location, file_name)
+                with FileSystems().create(target_path) as dst:
+                    shutil.copyfileobj(memfile, dst, WRITE_CHUNK_SIZE)
+
+                channel_names = [da.name for da in data]
+                tiff_data = TiffData(
+                    name=tiff_name,
+                    target_path=target_path,
+                    channel_names=channel_names,
+                    start_time=start_time,
+                    end_time=end_time,
+                    properties=attrs
+                )
+
+                yield tiff_data
+
+
+class IngestIntoEETransform(RateLimit):
+    """Ingests tiff image into earth engine and yields asset id.
+
+    Attributes:
+        ee_asset: The asset folder path in earth engine project where the tiff image files will be pushed.
+        ee_qps: Maximum queries per second allowed by EE for your project.
+        ee_latency: The expected latency per requests, in seconds.
+        ee_max_concurrent: Maximum concurrent api requests to EE allowed for your project.
+        private_key: A private key path to authenticate earth engine using private key. Default: None.
+        service_account: Service account address when using a private key for earth engine authentication.
+        use_personal_account: A flag to authenticate earth engine using personal account. Default: False.
+    """
+
+    def __init__(self,
+                 ee_asset: str,
+                 ee_qps: int,
+                 ee_latency: float,
+                 ee_max_concurrent: int,
+                 private_key: str,
+                 service_account: str,
+                 use_personal_account: bool):
+        """Sets up rate limit."""
+        self.ee_asset = ee_asset
+        super().__init__(global_rate_limit_qps=ee_qps,
+                         latency_per_request=ee_latency,
+                         max_concurrent_requests=ee_max_concurrent)
+        ee_initialize(use_personal_account=use_personal_account,
+                      service_account=service_account,
+                      private_key=private_key)
+
+    @retry.with_exponential_backoff(
+        num_retries=NUM_RETRIES,
+        logger=logger.warning,
+        initial_delay_secs=INITIAL_DELAY,
+        max_delay_secs=MAX_DELAY
+    )
+    def start_ingestion(self, asset_request: t.Dict) -> str:
+        """Creates COG-backed asset in earth engine. Returns the asset id."""
+        try:
+            result = ee.data.createAsset(asset_request)
+        except ee.EEException as err:
+            raise err
```

Comment:
I agree, I think the only reason to catch and then re-raise this exception would be to ensure logging happened. Given that the retry is in place, it may be that the first exceptions may not be logged as it retries.

Reply:
Yes, logging the error.